src/mbgl/annotation/shape_annotation_impl.cpp

```diff
@@ -19,84 +18,21 @@ ShapeAnnotationImpl::ShapeAnnotationImpl(const AnnotationID id_, const uint8_t m
-            converted.push_back(convertPoints(geojsonvt::ProjectedFeatureType::LineString, line));
-        }
-        return convertFeature(geojsonvt::ProjectedFeatureType::LineString, converted);
-    }
-
-    geojsonvt::ProjectedFeature operator()(const MultiPolygon<double>& polygons) const {
-        geojsonvt::ProjectedRings converted;
-        for (const auto& polygon : polygons) {
-            for (const auto& ring : polygon) {
-                converted.push_back(convertPoints(geojsonvt::ProjectedFeatureType::Polygon, ring));
-            }
-        }
-        return convertFeature(geojsonvt::ProjectedFeatureType::Polygon, converted);
-    }
-
-private:
-    geojsonvt::LonLat convertPoint(const Point<double>& p) const {
-        return {
-            util::wrap(p.x, -util::LONGITUDE_MAX, util::LONGITUDE_MAX),
-            util::clamp(p.y, -util::LATITUDE_MAX, util::LATITUDE_MAX)
-        };
-    }
-
-    geojsonvt::ProjectedRing convertPoints(geojsonvt::ProjectedFeatureType type, const std::vector<Point<double>>& points) const {
-        std::vector<geojsonvt::LonLat> converted;
-
-        for (const auto& p : points) {
-            converted.push_back(convertPoint(p));
-        }
-
-        assert(points.size() > 0);
-        if (type == geojsonvt::ProjectedFeatureType::Polygon && points.size() > 0 && points.front() != points.back()) {
-            converted.push_back(converted.front());
-        }
-
-        return geojsonvt::Convert::projectRing(converted, tolerance);
-    }
-
-    geojsonvt::ProjectedFeature convertFeature(geojsonvt::ProjectedFeatureType type, const geojsonvt::ProjectedRings& rings) const {
-        return geojsonvt::Convert::create(geojsonvt::Tags(), type, rings);
-    }
+struct ToGeometry {
+    mapbox::geometry::geometry<double> operator()(mapbox::geometry::line_string<double, std::vector> geom) const { return geom; }
+    mapbox::geometry::geometry<double> operator()(mapbox::geometry::polygon<double, std::vector> geom) const { return geom; }
+    mapbox::geometry::geometry<double> operator()(mapbox::geometry::multi_line_string<double, std::vector> geom) const { return geom; }
+    mapbox::geometry::geometry<double> operator()(mapbox::geometry::multi_polygon<double, std::vector> geom) const { return geom; }
 };
 
 void ShapeAnnotationImpl::updateTileData(const CanonicalTileID& tileID, AnnotationTileData& data) {
     static const double baseTolerance = 4;
 
     if (!shapeTiler) {
-        const uint64_t maxAmountOfTiles = 1 << maxZoom;
-        const double tolerance = baseTolerance / (maxAmountOfTiles * util::EXTENT);
-
-        std::vector<geojsonvt::ProjectedFeature> features = {
-            ShapeAnnotationGeometry::visit(geometry(), ToGeoJSONVT(tolerance))
-        };
-
+        ToGeometry toGeometry;
+        mapbox::geometry::feature_collection<double> features;
+        mapbox::geometry::feature<double> feature = { apply_visitor(toGeometry, geometry()) };
+        features.emplace_back(feature);
```

Comment:
I _think_ you should be able to write this using a generic lambda as the visitor:

```C++
FeatureCollection features;
features.emplace_back(ShapeAnnotationGeometry::visit(geometry(), [] (auto&& geom) {
    return Feature(geom);
}));
```

We don't have a `FeatureCollection` alias in `mbgl` but we should.